src/mapreduce.jl

```diff
@@ -235,7 +235,7 @@ end
 @generated function _diff(::Size{S}, a::StaticArray, ::Type{Val{D}}) where {S,D}
     N = length(S)
     Snew = ([n==D ? S[n]-1 : S[n] for n = 1:N]...)
-    T = Base.promote_op(-, eltype(a), eltype(a))
+    T = Core.Inference.return_type(-, Tuple{eltype(a),eltype(a)})
```

Comment:
IIRC calling `Inference.return_type` is considered a bit of a dirty trick by the core devs ;-) If used, it can change the behaviour of code depending on optimizations to the julia inference engine.  (Is this _the_ reason to avoid it?  Perhaps there's more and someone can enlighten me?).

I know we do it in another couple of places in `StaticArrays`, but I think we can avoid it here by doing something like

```julia
T = typeof(one(eltype(a)) - one(eltype(a)))
```

or some similar such trick?

The version for `_mapreducedim` further up is more tricky, because it takes an arbitrary `op` in the reduction. Unlike `diff`, it's not clear that the data will be numeric or that calling `one(eltype(a))` makes sense in that case.